refactor: Website routing and rendering

rmehta opened the pull request:
Website rendering is a mess. It's hard to debug and extend. This has been build incrementally without much focus on structure. There are multiple ways to resolve a route and the way the decision tree was handled is via exceptions. The code files were not well named and methods not grouped in any logical manner.

This PR tries to rebuild the website from scratch. Initial exploration shows that this can be a lot lot simpler than what it is. The approach will be to do test-driven development and try and then rebuild the pages. This module is being built in parallel and can be used via feature flag `flag_new_website`.

### New structure

Move to an object-oriented structure. Each type will be implemented in its own class:

```
StaticPage
TemplatePage
DocTypePage
WebFormPage (can be part of DocTypePage?)
DocsPage
```

Lots of features are redundant and can be removed. Each new feature must have a test.

### How you can contribute?

Complete feature coverage for each of the existing pages, starting by adding tests. Extend using these base classes.

### Tasks

- [x] Template pages (`/www`)
- [x] Static Files
- [x] Document pages (`/blog/<cat>/blog1`)
- [x] List pages (`/blog`)
- [x] Print pages
- [x] Documentation Pages
- [x] Redirect
- [x] Sitemap
- [x] Support `{"from_route": "/courses/<course>", "to_route": "courses/course?course=<course>"}` re-routing
- [x] Web Form
- [x] Sidebar
- [x] Fix caching
- [x] Fix custom homepage 404 caching
- [x] Update docs https://github.com/frappe/frappe_docs/pull/154

rmehta committed:
wip: refactor website routing and rendering

leela commented on frappe/tests/test_website_new.py:
```diff
@@ -0,0 +1,20 @@
+	def test_error_page(self):
+		set_request(method='GET', path='/error')
+		response = serve.TemplatePage().get()
+		self.assertEquals(response.status_code, 200)
```
Shouldn't this be 404 or 500?

rmehta replied:
yep. this is WIP - will fix!

leela commented on frappe/website/serve.py:
```diff
@@ -0,0 +1,276 @@
+
+	try:
+		path = path.strip('/ ')
+		resolve_redirect(path)
```
not cleat what it does until resolve_redirect method is seen. why not resolve_redirect returns a redirect path and use that here?

rmehta replied:
this is still legacy - trying to work on the core components first - then will handle the redirects

leela commented on frappe/website/serve.py:
```diff
@@ -0,0 +1,276 @@
+		path = resolve_path(path)
+		data = None
+
+		response = StaticPage(path).get()
+		if not response:
+			response = TemplatePage(path).get()
+		if not response:
+			response = DocTypePage(path).get()
+		if not response:
+			response = TemplatePage('404').get()
```
shouldn't we find what type of page it is based on path?

rmehta replied:
there are no rules that classify directly unless the route is resolved.

rmehta committed:
fix: status_code, login page works

rmehta committed:
fix(rendering): add doctype generate pages

rmehta committed:
fix(list): make list pages work

rmehta committed:
fix(minor): website/serve.py

rmehta committed:
fix(minor): fix some tests and merge old and new website tests

rmehta committed:
fix(minor): http_status_code

surajshetty3416 committed:
Merge branch 'develop' into refactor-website

surajshetty3416 committed:
feat: Option to consider query string while writing website redirect

- if match_with_query_string is true in a website redirect rule, system will try to match the regex with the result of route + query_string

surajshetty3416 committed:
refactor: Move page controllers to separate files

- Make template page work

surajshetty3416 committed:
refactor: Move code to appropriate file

add_sidebar_and_breadcrumbs belongs to TemplatePage

rmehta commented on frappe/website/page_controllers/template_page.py:
```diff
@@ -0,0 +1,287 @@
+			self.source = '''{{% extends "{0}" %}}
+				{{% block page_content %}}{1}{{% endblock %}}'''.format(context.base_template, self.source)
+
+		if "<!-- no-breadcrumbs -->" in self.source:
```
we can drop support for this - don't think anyone is using this anymore.

surajshetty3416 committed:
fix: Document Page

surajshetty3416 committed:
fix: Enable WebForm page

barredterra commented:
I made this quick diagram of class hierarchy to help me understand what's going on here. Hope it's useful for someone else as well:

![website](https://user-images.githubusercontent.com/14891507/112724596-674ab800-8f14-11eb-9c4c-0023a439fefb.png)

surajshetty3416 commented:
I'll re-open once this is ready for review.

surajshetty3416 committed:
Merge branch 'develop' into refactor-website

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
refactor: Remove support for commented attributes

- Use frontmatter instead

surajshetty3416 committed:
refactor: Simplify code
- Make code more readable

surajshetty3416 committed:
test: Use get_response instead of render in tests

surajshetty3416 committed:
fix: Replace render() with get_response

surajshetty3416 committed:
test: Fix Blog Post test case

surajshetty3416 committed:
fix: Add validator for WebForm

surajshetty3416 committed:
fix: set_metatags_from_website_route_meta

surajshetty3416 committed:
style: Fix code correctness issues

surajshetty3416 committed:
fix: Sitemap base template route

surajshetty3416 committed:
style: Resolve code correctness warning

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Base template path

surajshetty3416 committed:
refactor: Remove unnecessary code

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

barredterra commented on frappe/tests/test_website.py:
```diff
@@ -4,13 +4,19 @@
 
 import frappe
 from frappe.website import render
+from frappe.website import serve
```
```suggestion
from frappe.website.serve import get_response
```

The `serve.get_response()` below just sounds wrong. Like `serve` would contain a badly named _Server_ class. 😄

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
refactor: Import get_response instead of using serve.get_response

surajshetty3416 committed:
feat: Add class for standard pages (Error & NotFound)

surajshetty3416 committed:
refactor: Remove reduntant code
- Use Error & NotFound page as required

surajshetty3416 committed:
fix: Rename resolver to renderer

surajshetty3416 committed:
fix: Set basepath and toc data to show sidebar

surajshetty3416 committed:
fix: Template basepath

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
refactor: Delete unused code from render.py

surajshetty3416 committed:
fix: Update renderer sequence

surajshetty3416 committed:
fix: Circular import

surajshetty3416 committed:
refactor: Remove render method
- also, move clear cache to utils.py

surajshetty3416 committed:
refactor: Move metatags related code into a separate file

surajshetty3416 committed:
refactor: Move code to appropriate places

surajshetty3416 committed:
refactor: Remove render_page from render.py

& replace all usages of render_page with get_response

surajshetty3416 committed:
chore: Delete an unnecessary file
- purifycss is no longer used.

surajshetty3416 committed:
style: Fix import issues

surajshetty3416 committed:
fix: Add get_response_content to get page content

surajshetty3416 committed:
test: Use get_response_content to get content

surajshetty3416 committed:
refactor: Simplify sidebar and breadcrumbs code

surajshetty3416 committed:
refactor: Move unwanted code from validate

- Also, removed unwanted code in render.py

surajshetty3416 committed:
feat: Add redirect page

surajshetty3416 committed:
refactor: Add path_resolver to separate route resolving part

surajshetty3416 committed:
refactor: Move get_sidebar_data method to utils

- Also, rename get_sidebar_data to get_sidebar_items & delete context.py
- Remove commented text

surajshetty3416 committed:
style: Fix formatting issue

surajshetty3416 committed:
fix: Handle encoding issue and some edge cases

surajshetty3416 committed:
fix: Add cache_html decorator to cache HTML

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: 404 caching

surajshetty3416 committed:
test: Add test case to validate base_template_path

surajshetty3416 committed:
fix: 404 caching

- Move 404 resolving code to path resolver

surajshetty3416 committed:
fix: Do not set content type for paths ending with .com

To avoid unnecessary download

surajshetty3416 committed:
fix: Check if resolved file_path in StaticPage is a file

surajshetty3416 committed:
fix: Exception handling

surajshetty3416 committed:
feat: Add is_valid_path method to PathResolver

surajshetty3416 committed:
refactor: Delete unused code

surajshetty3416 committed:
refactor: Delete unused code

surajshetty3416 committed:
refactor: Move code to appropriate files

- Delete unused resolve_route method
- move resolve_path & resolve_redirect to path_resolver
- Remove redirect.py

surajshetty3416 committed:
Merge branch 'refactor-website' of github.com:rmehta/frappe into refactor-website

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
test: Add sidebar test

surajshetty3416 committed:
style: Translate string

surajshetty3416 committed:
fix: Add canonical link

surajshetty3416 committed:
test: Add test files

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Template path for document_page

surajshetty3416 committed:
fix: Website template path

surajshetty3416 committed:
fix: Fallback to template_path if template is not set

surajshetty3416 committed:
fix: Move set_missing_values to base_template_page

- Since it is common for all inherited pages from base_template_page

surajshetty3416 committed:
refactor: Rename validate function to can_render

surajshetty3416 committed:
fix: Make http_status_code non mandatory

surajshetty3416 committed:
fix: Add comment properties back with deprecation warning

surajshetty3416 committed:
refactor: Rename page_controllers folder to page_renderers

surajshetty3416 committed:
refactor: Remove render.py and move all utility functions to utils.py

- Replace or remove all render imports

surajshetty3416 committed:
test: Add test to check replacement of {next} & {index} identifiers

surajshetty3416 committed:
test: Check all cases of setting homepage

surajshetty3416 committed:
refactor: Use clear_website_cache instead of clear_cache

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
test: Clear hook cache after updating hook

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Colocated asset load

surajshetty3416 committed:
test: Add attributes via frontmatter

surajshetty3416 committed:
refactor: Move context value setting code to a separate function

surajshetty3416 committed:
test: Use assertIn for better error message

surajshetty3416 committed:
fix: Init page properties in webpage

surajshetty3416 committed:
fix: Breadcrumbs for a page

surajshetty3416 committed:
test: Add test case to check breadcrumbs

surajshetty3416 committed:
test: Refactor webpage test to remove unnecessary operations

surajshetty3416 committed:
fix: get_index_path_options method

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Set default value as empty string

surajshetty3416 committed:
fix: Blog post style

surajshetty3416 committed:
fix: Deprecation warning

surajshetty3416 committed:
feat: Option to add custom renderers

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Custom page renderer logic

surajshetty3416 committed:
test: Add test case to validate custom page renderer

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Remove unnecessary import

surajshetty3416 committed:
Merge branch 'develop' into refactor-website

surajshetty3416 committed:
test: Add printview page test

surajshetty3416 committed:
Merge branch 'refactor-website' of github.com:rmehta/frappe into refactor-website

surajshetty3416 committed:
fix: WebForm render

surajshetty3416 committed:
test: Simple test to validate WebForm rendering

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Rename WebPage to BaseRenderer

surajshetty3416 committed:
Merge branch 'develop' of https://github.com/frappe/frappe into refactor-website

surajshetty3416 committed:
fix: Only replace hyphen in file name

surajshetty3416 committed:
fix: Remove deprecated comments

surajshetty3416 committed:
chore: debug

netchampfaris commented:
@surajshetty3416 We need to publish a migration guide for upgrading from older version.

surajshetty3416 commented:
## Migration Guide

We have moved few methods to a different file. Please make the following changes if your app is breaking after this change.

```diff
- from frappe.website.render import render
+ from frappe.website.serve import get_response
...
- response = render()
+ response = get_response()
```

```diff
- from frappe.website.render import clear_cache
+ from frappe.website.utils import clear_cache
```

https://github.com/frappe/frappe/wiki/Migrating-to-Version-14
